api/api-plugin-infos-v5/src/main/java/com/thoughtworks/go/apiv5/plugininfos/PluginInfosControllerV5.java

```diff
@@ -0,0 +1,172 @@
+        super(ApiVersion.v5);
+        this.apiAuthenticationHelper = apiAuthenticationHelper;
+        this.pluginInfoFinder = pluginInfoFinder;
+        this.entityHashingService = entityHashingService;
+        this.defaultPluginManager = defaultPluginManager;
+    }
+
+    @Override
+    public String controllerBasePath() {
+        return Routes.PluginInfoAPI.BASE;
+    }
+
+    @Override
+    public void setupRoutes() {
+        path(controllerBasePath(), () -> {
+            before("", mimeType, this::setContentType);
+            before("/*", mimeType, this::setContentType);
+
+            before("", this.mimeType, this.apiAuthenticationHelper::checkAdminUserAnd403);
+            before(Routes.PluginInfoAPI.ID, this.mimeType, this.apiAuthenticationHelper::checkAdminUserAnd403);
+
+            get(Routes.PluginInfoAPI.ID, mimeType, this::show);
+            get("", mimeType, this::index);
+
+            exception(HttpException.class, this::httpException);
+        });
+    }
+
+    public String index(Request request, Response response) throws IOException {
+        List<CombinedPluginInfo> pluginInfos = new ArrayList<>();
+        String pluginType = request.queryParams("type");
+        Boolean includeBad = Boolean.valueOf(request.queryParams("include_bad"));
+
+        try {
+            Collection<CombinedPluginInfo> validPluginInfos = this.pluginInfoFinder.allPluginInfos(pluginType).stream()
+                    .filter(pluginInfo -> !hasUnsupportedExtensionType(pluginInfo))
+                    .collect(Collectors.toList());
+
+            pluginInfos.addAll(validPluginInfos);
+        } catch (InvalidPluginTypeException exception) {
+            throw new UnprocessableEntityException(String.format("Invalid plugins type - `%s` !", pluginType));
+        }
+
+        if (includeBad) {
+            List<BadPluginInfo> badPluginInfos = defaultPluginManager.plugins().stream()
+                    .filter(GoPluginDescriptor::isInvalid)
+                    .map(BadPluginInfo::new)
+                    .collect(toList());
+
+            pluginInfos.addAll(badPluginInfos);
+        }
+
+        String etag = etagFor(pluginInfos);
+
+        if (fresh(request, etag)) {
+            return notModified(response);
+        }
+        setEtagHeader(response, etag);
+        return writerForTopLevelObject(request, response, writer -> PluginInfosRepresenter.toJSON(writer, pluginInfos));
+
+    }
+
+    public String show(Request request, Response response) throws IOException {
+        String pluginId = request.params("id");
+        CombinedPluginInfo pluginInfo = this.pluginInfoFinder.pluginInfoFor(pluginId);
+        CombinedPluginInfo badPluginInfo = null;
+
+
+        if (pluginInfo != null && hasUnsupportedExtensionType(pluginInfo)) {
```

Comment:
Can be simplified

```
case1:   pluginInfo === null => return RecoredNotFound
case2:  isBadPluignInfo => return response(instead of record not found)
else return pluginInfo




```